### DISCUSSION TITLE
Webauthn level support

### DISCUSSION THREAD

**Discussion by daniel-mezo** [webauthn4j/webauthn4j#532] [Q&A] [answered]
Hi guys,

I am looking to understand what level of Webauthn this library supports and will intend to support in the future.

The assertion verification during the authentication ceremony as implemented in [AuthenticationDataValidator](https://github.com/webauthn4j/webauthn4j/blob/0.16.1.RELEASE/webauthn4j-core/src/main/java/com/webauthn4j/validator/AuthenticationDataValidator.java#L69) indicates [Webauthn level 1](https://www.w3.org/TR/webauthn-1/#verifying-assertion), reflected by the 18 commented steps matching to **that** specification, but not to newer ones, where this ceremony involves more steps actually.
- Webauthn level 2: 22 steps, see https://www.w3.org/TR/webauthn-2/#sctn-verifying-assertion
- Webauthn level 3: 22 steps, see https://www.w3.org/TR/webauthn-3/#sctn-verifying-assertion

But if I look at data classes, there are also examples of the library implementing entries such as the [Enterprise Attestation Conveyance Preference](https://github.com/webauthn4j/webauthn4j/blob/master/webauthn4j-core/src/main/java/com/webauthn4j/data/AttestationConveyancePreference.java#L68), which only have been introduced in newer Webauthn levels, in this case this preference was added by [Webauthn level 2](https://services.w3.org/htmldiff?doc1=https%3A%2F%2Fwww.w3.org%2FTR%2Fwebauthn-1%2F&doc2=https%3A%2F%2Fwww.w3.org%2FTR%2Fwebauthn-2%2F#enum-attestation-convey).

Furthermore, there have been a very recent draft release of [Webauthn level 3](https://w3c.github.io/webauthn/), which obviously I wouldn't expect the library or authenticators to support yet, but it makes it clear that the specification is under active development.

So my questions:
- What level of Webauthn is currently supported by the library? Is it a mix of levels? Are these levels backward compatible (It does not seem to be written into the spec as far as I see)?
- Will newer levels of Webauthn be integrated into the library consistantly?
- While we are at it, will the new non-backward compatible [MD3](https://medium.com/webauthnworks/webauthn-fido2-whats-new-in-mds3-migrating-from-mds2-to-mds3-a271d82cb774) be integrated into the library?

**Comment by ynojima** [comment, marked as answer]
> What level of Webauthn is currently supported by the library?

As of today, this library supports WebAuthn level1 verification, but the data classes are able to parse level2 data.
As the level2 became w3c recommendation recently, I'd like to add level2 verification support in the near future.

> Will newer levels of Webauthn be integrated into the library consistantly?

It depends on the case, but when it become w3c recommendation, I'll start implementation.

> While we are at it, will the new non-backward compatible MD3 be integrated into the library?

MD3 support will be added in the future, but new "webauthn4j-metadata" module will be implemented in backward-incompatible way.

**Reply by daniel-mezo** [reply]
Thank you for the response @ynojima!

**Reply by daniel-kummer** [reply]
I'd also like to thank you @ynojima 👍

**Reply by ynojima** [reply]
[0.17.0.RELEASE](https://github.com/webauthn4j/webauthn4j/releases/tag/0.17.0.RELEASE) with Level2 support is released today.

**Comment by ynojima** [comment]
Tracking issue is created: https://github.com/webauthn4j/webauthn4j/issues/533
